src/types.rs

```diff
@@ -48,11 +96,23 @@ impl TypeRef {
                 BuiltInTypeDefinition::Derived(other_type) => other_type.name().to_owned().unwrap(),
             },
         };
+
+        // convert given IonSchemaElement to an OwnedElement
+        // converts document type to SExpression
+        let value = match value.into() {
+            IonSchemaElement::Element(element) => element,
+            IonSchemaElement::Document(document) => {
+                OwnedElement::new_sexp(document.iter().cloned())
+                    .with_annotations(vec![text_token("document")])
+            }
```

Comment:
issue—You are using `document::` as a meta-annotation, but it's completely valid for someone to have `document::(/*...*/)` within regular Ion data. You need some way to [escape](https://en.wikipedia.org/wiki/Escape_character) a "regular" usage of `document::(/*...*/)`. However, anything that you could use as an escape marker is also valid Ion, and would also need to be escaped. There are a few solutions to this:

1. Escape "regular" uses of `document::` with `#document::` or similar. This means that you also need to escape "regular" uses of `#document::` as `##document::` and so on.
2. Wrap all `document` _and_ `sexp` in another `sexp`. Eg. the document `a b c` becomes `document::(a b c)` and the sexp `(a b c)` becomes `sexp::((a b c))`.
3. Don't try to encode a document as a single Ion value. Document _can_ be a hard-coded type, even though it's not an Ion type.

I think option 3 will be the simplest in the long run, even if it means doing some significant refactoring right now. It's also the only option that doesn't start adding in a bunch of computational (and memory allocation) overhead for wrapping and unwrapping the documents as sexps.

Reply:
I think choosing option 3 like you said would need some significant refactoring including having a check for `document` type for each `validate` methods. But I do think this option is better and moves away from all these tricks or `meta-annotation` issues. I will try to refactor using option 3.

Reply:
I have now replaced all places where converting document to sexpression. Now `IonSchemaElement` is used everywhere in the validation logic.